draft-bbs-signatures.md

```diff
@@ -291,6 +291,63 @@ For readability, this document makes these transformations implicitly, but they
 
 Optional input/parameters to operations that feature in a call to a HASH or XOF function, that are not supplied to the operation should default to an empty octet string. For example, if X is an optional input/parameter that is not supplied, whilst A and B are required, then the procedural step of `HASH(A || X || B)` MUST be evaluated to `HASH(A || "" || B)`.
 
+## Generator point computation
+
+The `compute_generators` function defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify. It takes one input, the maximum number of selectively revealed messages, which determines the number of generator points to create. The result is a sequence beginning with the `H_s` and `H_d` generators, followed by each of the `H_i` message generators.
+
+This document defines one implementation of this method, `compute_generators_xof`, which uses [@!I-D.irtf-cfrg-hash-to-curve] with an extensible-output hash function to generate a sequence of points with a uniformly random distribution from a seed value and DST.
+
+Each generator creation method must define a unique `GENERATORS_ID` identifier which is incorporated into the ciphersuite ID.
+
+### Security Considerations
+
+TODO
+
+### compute\_generators\_xof
+
+```
+generators = compute_generators_xof(maxMessageCount);
+
+Parameters:
+
+- hash_to_curve_g1: A cryptographic hash function that takes as an arbitrary octet string input and returns a point in G1 as defined in [@!I-D.irtf-cfrg-hash-to-curve]. The expand_message method MUST be expand_message_xof, and hash MUST be the same as xof.
+- generator_xof, an extensible-output cryptographic hash function.
+- generator_dst, octet string. The domain separation tag.
+- generator_seed, octet string. A unique seed value for the generator point sequence.
+
+Inputs:
+
+- maxMessageCount (REQUIRED), non-negative integer. The maximum number of selectively-revealed messages to be supported.
```

Comment:
```suggestion
- maxMessageCount (REQUIRED), non-negative integer. The number of generators to be returned.
```

If the operation is to be called inside each procedure, should this be something like the above?? Otherwise if this input is indeed the maximum number of messages that an issuer can (ever) sign, dosen't this dissallow what the previous version of the spec did (i.e., something like `g_i = hash_to_curve(PK || i || L)`) because in this case it will be weistfull to create maxMessageCount generators??

Reply:
It's meant to be the maximum number of messages supported by this set of generators, but yes that could be clearer. Having a dependency on L in the seed seems like a liability to me.

Reply:
I agree regarding the seed and L dependency. I would much rather say that the input is a ciphersuite-defined seed. So we don't have to worry a lot about adversarial inputs which can be a pain to deal with (even if we use an XOF only).

IMO there are the following trade-offs in the different input choices to `compute_generators` :

- If we use maxMessageCount, the generators can be a parameter and not created in the procedure of each operation. This makes it clearer that they should be cached.
- On the cons side, it "forces" a hard upper limit to the number of messages that can be signed, without any "technical" reason. It also requires the users to define this number which may not be as "straight forward", since, the more messages they "support", the less efficient the operation will be and the more generators they will need to cache.
- Calling `compute_generators(L)` inside each procedure is clearer IMO and doesn't require more "work" from the implementer to define parameters.
- On the cons side, it may not be clear enough that the output of `compute_generators` should be cached and **NOT** be re-calculated each time.

I would personally go with option 2 and call `compute_generators(L)` inside each operation, with a note that the output must be cached (if it's not, it is not a security issue, just an efficiency one, so i think it's fine).

Reply:
+1. In this case i would suggest `maxMessageCount` -> `MessageCount` (or `message_count`). I think this will be clearer.

Reply:
Note, in this case `L` should be inputted to `ProofVerify`. Not really an issue (i know a lot of implementations do this already), just a note so i will not forget to update the inputs.